client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java

```diff
@@ -168,10 +168,15 @@ private boolean dataWaitingToGoInBuffer() {
     }
 
     private void handleRequest() throws SegmentTruncatedException {
-        if (outstandingRequest.isCompletedExceptionally()) {
-            issueRequestIfNeeded();
+        SegmentRead segmentRead;
+        try {
+            segmentRead = outstandingRequest.join();
+        } catch (Exception e) {
+            if (Exceptions.unwrap(e) instanceof SegmentTruncatedException) {
+                receivedTruncated = true;
+            }
+            throw e;
```

Comment:
The exceptions that can be thrown by outstandingRequest.join() are 
1. RetriesExhaustedException (This will be thrown by EventStreamReader.readNextEvent(timeout))
2. ConnectionClosedException. (this will never happen in the case here.)
3. SegmentTruncatedException (this is already handled in the code.)


`SegmentInputStreamImpl#issueRequestIfNeeded` currently reissues a read if `outstandingRequest `completes exceptionally and if the exception is not an `ERROR `or `InterruptedException `or `CancellationException`.

1) We could Modify IssueRequestIfNeeded to handle RetriesExhaustedExeption too right?
2) Why even have this check since AysncSegmentInputStream.read() will not have any other exceptions apart from the 3 listed above.

Reply:
Excellent points. By handling this in handelRequest() we can simplify issueRequestIfNeeded().
I've updated the code to reflect this.